classes/observation_manager.php

```diff
@@ -296,4 +298,96 @@ public static function reorder_observation_point(int $observationid, int $obpoin
+     * @param int $observationid ID of the observation instance
+     * @param int $sessionid ID of the observation sesssion
+     * @return array array of observation points with responses for the given session
+     */
+    public static function get_points_and_responses(int $observationid, int $sessionid) {
+        global $DB;
+
+        // Selects all points for this observation,
+        // and attaches responses for the current session (if one exists).
+
+        $sql = 'SELECT pts.id as point_id, obs_id, title, list_order, ins, ins_f, max_grade, res_type,
+        sess_resp.id as response_id, obs_ses_id as session_id, grade_given, response, ex_comment
+        FROM {observation_points} pts
+        LEFT JOIN
+        (SELECT * FROM {observation_point_responses} resp WHERE obs_ses_id = :sessionid) as sess_resp
+        ON pts.id = sess_resp.obs_pt_id WHERE pts.obs_id = :observationid;';
+
+        $pointsandresponses = $DB->get_records_sql($sql, ['observationid' => $observationid, 'sessionid' => $sessionid]);
+
+        return $pointsandresponses;
+    }
+
+    /**
+     * Submits a response to a particular observation point for a given sesssion.
+     * @param int $sessionid ID of the observation session
+     * @param int $pointid ID of the observation point this response is for
+     * @param mixed $data data object returned from the pointmarking_form
+     */
+    public static function submit_point_response(int $sessionid, int $pointid, $data) {
+        global $DB;
+
+        if ($data->grade_given < 0 || !is_int($data->grade_given)) {
+            throw new \coding_exception("Grade given must be an integer that is zero or more");
+        }
+
+        if (!isset($data->response)) {
+            throw new \coding_exception("No response was found in the data.");
+        }
+
+        $sessioninfo = \mod_observation\session_manager::get_session_info($sessionid);
+        $pointdata = self::get_existing_point_data($sessioninfo['obid'], $pointid);
+        if ($data->grade_given > $pointdata->max_grade) {
+            throw new \coding_exception("Grade given must be less than the max grade.");
+        }
+
+        // See if a response already exists for this session and pointid.
+        $existingresponse = $DB->get_record('observation_point_responses', ['obs_pt_id' => $pointid, 'obs_ses_id' => $sessionid]);
+
+        // Clean data.
+        $dbdata = [
+            'obs_pt_id' => $pointid,
+            'obs_ses_id' => $sessionid,
+            'grade_given' => $data->grade_given,
+            'response' => $data->response,
+            'ex_comment' => $data->ex_comment,
+        ];
+
+        if ($existingresponse === false) {
+            // Insert new.
+            $DB->insert_record('observation_point_responses', $dbdata);
+        } else {
+            // Update existing.
+            $dbdata['id'] = $existingresponse->id;
+            $DB->update_record('observation_point_responses', $dbdata);
```

Comment:
generally for items like this, its best practice to have an additional 2 DB fields, `timecreated` and `timemodified`

so here you could set both to `time()` for new inserts, or just touch timemodified for updates